src/pages/experiments/[experimentId]/data-processing/configure-embedding/components/EmbeddingPreview.jsx

```diff
@@ -81,23 +78,55 @@ const EmbeddingPreview = () => {
     },
   };
 
+  // Start loading by fetching the settings and cell sets.
+  useEffect(() => {
+    dispatch(loadCellSets(experimentId));
+    dispatch(loadProcessingSettings(experimentId, FILTER_UUID));
+  }, []);
+
+  // Once the config is loaded, we can also load the embedding.
+  useEffect(() => {
+    if (!embeddingMethod) {
+      return;
+    }
+
+    dispatch(loadEmbedding(experimentId, embeddingMethod));
+  }, [embeddingMethod]);
+
   useEffect(() => {
     // Do not update anything if the cell sets are stil loading or if
     // the config does not exist yet.
-
     if (!config) {
       return;
     }
 
+    // If no embedding method is available yet, the config did not load.
+    // We can skip processing until it is done.
+    if (!embeddingMethod) {
+      return;
+    }
+
+    // If the config is loaded, but the embedding is still loading, we
+    // don't have any data to add.
+    if (!embedding[embeddingMethod] || embedding[embeddingMethod].loading) {
+      return;
+    }
+
+    // If cell sets don't exist yet or they're loading, wait until they are done.
+    if (!cellSets || cellSets.loading) {
+      return;
+    }
+
+    // Otherwise, begin by creating the spec appropriate for the plot.
     const spec = plots[selectedSpec].specGenerator(config);
     generateData(spec);
     setPlotSpec(spec);
-  }, [config]);
+  }, [config, embeddingSettings, embedding]);
 
-  // If the user toggles to a different embedding, set the config to be the initial
-  // state for that type of plot.
   useEffect(() => {
-    setConfig(plots[selectedSpec].initialConfig);
+    // If the user toggles to a different embedding, set the config to be the initial
+    // state for that type of plot.
+    setConfig(_.cloneDeep(plots[selectedSpec].initialConfig));
   }, [selectedSpec]);
 
   // Quick and dirty function to massage prepared data into a good shape.
```

Comment:
@aerlaut you are deleting this function as part of your PR. Is this change in here going to clash very badly with your changes?

Reply:
Yes. with the refactor, `embeddingPreview` becomes very trimmed down. The latest changes can be seen here : https://github.com/biomage-ltd/ui/pull/95/files#diff-afd1e0b0de5b6da0610c8bd7fbb9865fec7ffc38aaf347805381ff303ea967bb

Reply:
Correct, although I can see that you basically implemented a large chunk of the ticket that I was assigned as well -- this makes sense given your scope I guess, and I can see that we can overwrite most of my changes with this of yours. That said, I think we should merge yours after mine to make sure that nothing that I verified as working as part of my ticket was somehow left out by yours.